crates/aptos-in-memory-cache/src/caches/fifo.rs

```diff
@@ -0,0 +1,426 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Cache, OrderedCache, StreamableOrderedCache};
+use dashmap::DashMap;
+use futures::{stream, Stream};
+use parking_lot::RwLock;
+use std::{fmt::Debug, hash::Hash, sync::Arc};
+use tokio::sync::Notify;
+use tokio_util::sync::CancellationToken;
+use tracing::info;
+
+#[derive(Debug, Clone, Copy)]
+struct CacheMetadata<K> {
+    eviction_trigger_size_in_bytes: u64,
+    target_size_in_bytes: u64,
+    total_size_in_bytes: u64,
+    last_key: Option<K>,
+    first_key: Option<K>,
+}
+
+/// A simple in-memory cache with a deterministic FIFO eviction policy.
+pub struct FIFOCache<K, V>
+where
+    K: Hash + Eq + PartialEq + Send + Sync + Clone + 'static,
+    V: Send + Sync + Clone + 'static,
+{
+    /// Precache to support out of order inserts
+    precache: Arc<DashMap<K, V>>,
+    precache_insert_notify: Arc<Notify>,
+    /// Cache maps the cache key to the deserialized Transaction.
+    items: Arc<DashMap<K, V>>,
+    insert_notify: Arc<Notify>,
+    cache_metadata: Arc<RwLock<CacheMetadata<K>>>,
+    _cancellation_token_drop_guard: tokio_util::sync::DropGuard,
+    // User defined function to get the next key for a given key
+    // The function provides the key and the getter function to the DashMap in case a lookup is necessary
+    next_key_function:
+        Arc<dyn Fn(&K, &dyn Fn(&K) -> Option<V>) -> Option<K> + Send + Sync + 'static>,
+}
+
+impl<K, V> FIFOCache<K, V>
+where
+    K: Debug + Hash + Eq + PartialEq + Send + Sync + Clone + 'static,
+    V: Send + Sync + Clone + 'static,
+{
+    pub fn new(
+        target_size_in_bytes: u64,
+        eviction_trigger_size_in_bytes: u64,
+        next_key_function: impl Fn(&K, &dyn Fn(&K) -> Option<V>) -> Option<K> + Send + Sync + 'static,
+    ) -> Self {
+        let cancellation_token: CancellationToken = CancellationToken::new();
+        let precache = Arc::new(DashMap::new());
+        let precache_insert_notify = Arc::new(Notify::new());
+        let items = Arc::new(DashMap::new());
+        let insert_notify = Arc::new(Notify::new());
+        let cache_metadata = Arc::new(RwLock::new(CacheMetadata {
+            eviction_trigger_size_in_bytes,
+            target_size_in_bytes,
+            total_size_in_bytes: 0,
+            last_key: None,
+            first_key: None,
+        }));
+        let next_key_function = Arc::new(next_key_function);
+
+        let cache = Self {
+            precache,
+            precache_insert_notify,
+            items,
+            insert_notify,
+            cache_metadata,
+            _cancellation_token_drop_guard: cancellation_token.clone().drop_guard(),
+            next_key_function,
+        };
+
+        cache.spawn_insertion_task(cancellation_token.clone());
+        cache.spawn_eviction_task(cancellation_token);
```

Comment:
Should we make a struct to represent these 2 threads? Otherwise they're just floating around. Imagine you need to instantiate multiple of these cache, it'd be hard to track / log the specific threads that work with a particular cache. 

My suggestion would be to have a CacheThread object, and add it to the cache. See if it's possible to generalize the insertion and eviction task as well so that you can basically instantiate the CacheThread with the task.